server/RdtClient.Service/Services/TrackerListGrabber.cs

```
@@ -0,0 +1,165 @@
+
+        using (response)
+        {
+            response.EnsureSuccessStatusCode();
+
+            await using var contentStream = await response.Content.ReadAsStreamAsync(token).ConfigureAwait(false);
+            using var reader = new StreamReader(contentStream);
+            var result = await reader.ReadToEndAsync(token).ConfigureAwait(false);
+
+            String[] trackers;
+
+            try
+            {
+                trackers = result
+                           .Split([
+                                      "\r\n", "\n"
+                                  ],
+                                  StringSplitOptions.RemoveEmptyEntries)
+                           .Where(line => !String.IsNullOrWhiteSpace(line) && !line.TrimStart().StartsWith('#'))
+                           .Select(t => t.EndsWith("/") ? t.TrimEnd('/') : t)
+                           .Select(t => t.Trim())
+                           .Where(t =>
+                           {
+                               if (!Uri.TryCreate(t, UriKind.Absolute, out var uri))
+                               {
+                                   return false;
+                               }
+
+                               return (uri.Scheme == Uri.UriSchemeHttp ||
+                                       uri.Scheme == Uri.UriSchemeHttps ||
+                                       uri.Scheme.Equals("udp", StringComparison.OrdinalIgnoreCase) ||
+                                       uri.Scheme.Equals("wss", StringComparison.OrdinalIgnoreCase)) &&
+                                      !t.Contains("..") &&
+                                      !t.Contains("\\") &&
+                                      !t.Any(Char.IsControl) &&
+                                      uri.Host.All(c => Char.IsLetterOrDigit(c) || c == '.' || c == '-' || c == '_' || c == ':') &&
+                                      uri.Host.Length > 0;
```

I'm undecided whether it's worth logging when we reject a tracker.
On the one hand, it would be useful for users to be able to see if/why a tracker is rejected, rather than `rdt-client` doing it silently.
On the other hand, that's potentially going to be incredibly verbose, and will happen every time we get the tracker list (which is only every hour by default)

---

I'm torn about this due to the pros and cons you outlined. If they were to be logged, it's probably best to do so only at a debug level. We could then log at a different level, such as Information, a simple count.

`logger.LogDebug("Rejected tracker: {TrackerUrl} - Reason: Invalid format or unsupported scheme.", t);`
`logger.LogInformation("{TrackerRejectionCount} trackers were rejected during enrichment.", trackerRejectionCount);`

Maybe?

---

I think his is a reasonable balance - users can get some information at `Information`, and we can see *why* when we ask for `Debug` logs